File: node.cpp
#include <zmq.hpp>
#include <string>
#include <iostream>
#include <vector>
#include <cstdlib>
#include <unistd.h>

using namespace std;

const string clientPrefix="tcp://localhost:";
const string serverPrefix="tcp://*:";
const int dataSize=9;
const string genericData="publicKey";

int procNumber, totalProcs;
int basePort=8000;
int broadcastPort=7000;
zmq::socket_t *reqSocket, *repSocket;
zmq::socket_t **broadSockets;

string s_recv (zmq::socket_t *socket) {

    zmq::message_t request;
    socket->recv(&request);
    return string(static_cast<char*>(request.data()), request.size());
}

void s_send (zmq::socket_t *socket, string send) {

    zmq::message_t request(send.length());
    memcpy(request.data(), send.c_str(), send.length());
    socket->send(request);
}

vector<int> getAddShares(int in)  
{
    string resp;
    if(procNumber) {
        cout<<"Received "<<s_recv(broadSockets[0])<<endl;

        resp = s_recv(repSocket,genericData);
        // multiply encrypted input by waht we just got
        s_send(reqSocket, genericData);
        resp = s_recv(reqSocket)

        //select a random share 
        s_send(broadSockets[0],genericData); //send share back to p1
        s_send(repSocket, genericData); //send something backwards

        resp = s_recv(broadSockets[0]); 

    } else {
        //send public key e to all parties
        for(int i=0; i<(totalProcs-1); i++) {
            s_send(broadSockets[i],genericData);
        }

        s_send(reqSocket, genericData); //send encrypted x1

        for(int i=0; i<(totalProcs-1); i++) {
            resp=s_recv(broadSockets[i]); // collect all the shares
        }

        resp = s_recv(reqSocket); 
        //now compute its own share. We know all shares now

        for(int i=0; i<(totalProcs-1); i++) {
            s_send(broadSockets[i],genericData); //send all shares
            //resp=s_recv(broadSockets[i]); // receive acknowledgement
        }
    }
    return vector<int>();
}

int main (int argc, char* argv[])
{
    procNumber = atoi(argv[1]);
    totalProcs =atoi(argv[2]);
    if(argc>3) {
        basePort=atoi(argv[3]);
        broadcastPort=basePort+10;
    }

    cout<<"Starting "<<procNumber<<" "<<totalProcs<<endl;

    zmq::context_t context (1);
    reqSocket = new zmq::socket_t (context, ZMQ_REQ);
    repSocket = new zmq::socket_t (context, ZMQ_REP);

    //cout<<clientPrefix+to_string(basePort+procNumber+1)<<" "<<serverPrefix+to_string(basePort+procNumber)<<endl;

    reqSocket->connect(clientPrefix+to_string(basePort+procNumber+1));
    repSocket->bind(serverPrefix+to_string(basePort+procNumber));


    if(procNumber) {
        broadSockets = new zmq::socket_t*[1];
        broadSockets[0] = new zmq::socket_t (context, ZMQ_REP);
        broadSockets[0]->bind(serverPrefix+to_string(broadcastPort+procNumber));
    } else {
        broadSockets = new zmq::socket_t*[totalProcs-1];
        for(int i=1;i<totalProcs;i++) {
            broadSockets[i-1] = new zmq::socket_t (context, ZMQ_REQ);
            broadSockets[i-1]->connect(clientPrefix+to_string(broadcastPort+i));
        }
    }

    usleep(2000000);
    getAddShares(0);

/*
    reqSocket->close();
    repSocket->unbind(serverPrefix+to_string(basePort+procNumber));
    if(procNumber) broadSockets[0]->unbind(serverPrefix+to_string(broadcastPort+procNumber));
    else for(int i=1;i<totalProcs;i++) broadSockets[i-1]->close();
    */

    return 0;
} 

File: node.py
import time
import sys
import zmq

from getAddShares import *
#from paillier.paillier import *
from modular import *
from random import *
from means import *
from datapoint import *

clientPrefix="tcp://localhost:";
serverPrefix="tcp://*:";
dataSize=9;
genericData="publicKey";
clusters = 4

basePort = 8000
broadcastPort = 9000
broadSockets=[]
leaderSocket = None
verbose = False

procNumber = int(sys.argv[1])
totalProcs = int(sys.argv[2])
clusters = int(sys.argv[3])
iters = int(sys.argv[4])
if len(sys.argv)>3:
    basePort = int(sys.argv[5])
    broadcastPort = basePort + 100

context = zmq.Context()
requestSocket = context.socket(zmq.REQ)
replySocket = context.socket(zmq.REP)
kmeans = KMeans(clusters, "data/horizontal-k-means/"+str(procNumber)+".in")
dimensions = kmeans.getDims()

requestSocket.connect(clientPrefix + str(basePort+procNumber+1))
replySocket.bind(serverPrefix + str(basePort+procNumber))

if procNumber > 0:
    leaderSocket = context.socket(zmq.REP)
    leaderSocket.bind(serverPrefix + str(broadcastPort+procNumber))
    broadSockets = None
else:
    for i in range(1,totalProcs):
        broadSockets.append(context.socket(zmq.REQ))
        broadSockets[i-1].connect(clientPrefix + str(broadcastPort+i))

timing = [time.time()]
for iteration in range(iters): # iterate 20 times for now
    centroids = kmeans.getClusterCentroids()
    newMeans = []
    for centroid, num in centroids:
        denom = getAddShares(procNumber, totalProcs, num, leaderSocket, replySocket, requestSocket, broadSockets)
        if denom == 0 :
            newMeans.append(centroid)
            continue
        newVal = []
        val = centroid.getVector()
        for i in range(dimensions):
            num = getAddShares(procNumber, totalProcs, val[i], leaderSocket, replySocket, requestSocket, broadSockets)
            newVal.append(num / (1.0 * denom))
        newMeans.append(DataPoint(dimensions, newVal))
    if verbose and not procNumber:
        print "[Process", str(procNumber)+"] iteration", iteration, ":", kmeans
    kmeans.updateMeans(newMeans)
    timing.append(time.time())

if not procNumber:
    results = [float("%0.2f" % (timing[i+1]-timing[i])) for i in range(len(timing)-1)]
    print results
    print sorted(results)[len(results)/2]

# requestSocket.unbind()
# replySocket.disconnect()
# if procNumber > 0:
#     broadSockets[0].unbind()
# else:
#     for i in range(1,totalProcs):
#         broadSockets[i-1].disconnect()
